montblanc/impl/rime/tensorflow/rime_ops/zernike_op_gpu.cuh

```diff
@@ -149,33 +164,61 @@ __global__ void zernike_dde_zernike(
     int ant = blockIdx.y * blockDim.y + threadIdx.y;
     int time = blockIdx.z * blockDim.z + threadIdx.z;
 
-  int i = threadIdx.z * 4 + (threadIdx.x & 0x03);
+    if(corr >= 4 || chan >= nchan || ant >= na || time >= ntime) return;
 
-    if(i < npoly){  
-        shared.zernike_coeff[threadIdx.y][threadIdx.x >> 2][i] = in_coeffs[((ant * nchan + chan) * npoly + i) * 4 + corr];
-        shared.zernike_noll_indices[threadIdx.y][threadIdx.x >> 2][i] = in_noll_index[((ant * nchan + chan) * npoly + i) * 4 + corr];
+    if(threadIdx.z < npoly)
+    {
+        shared.zernike_coeff[threadIdx.y][threadIdx.x >> 2][threadIdx.z] = in_coeffs[
+            ((ant * nchan + chan) * npoly + threadIdx.z) * 4 + corr];
+        shared.zernike_noll_indices[threadIdx.y][threadIdx.x >> 2][threadIdx.z] = in_noll_index[
+            ((ant * nchan + chan) * npoly + threadIdx.z) * 4 + corr];
     }
         
-    if(threadIdx.z == 0) shared.antenna_scaling[threadIdx.y][threadIdx.x >> 2] = in_antenna_scaling[(((ant * nchan + chan) * 4 + corr))];
-    if((threadIdx.x & 0x03) == 0){
-        shared.pointing_error[threadIdx.z][threadIdx.y][threadIdx.x>>2]  = in_pointing_error[(((time * na + ant) * nchan + chan) * 4 + corr)];
+    if(threadIdx.z == 0)
+    {
+        shared.antenna_scaling[threadIdx.y][threadIdx.x >> 2] = in_antenna_scaling[
+            (((ant * nchan + chan)))];
+    }
+    if((threadIdx.x & 0x03) == 0)
+    {
+        shared.pointing_error[threadIdx.z][threadIdx.y][threadIdx.x>>2]  = in_pointing_error[
+                (((time * na + ant) * nchan + chan))];
+    }
 
+    if(threadIdx.x == 0){
+        shared.pa_sin[threadIdx.z][threadIdx.y] = in_parallactic_angle_sin[time * na + ant];
+        shared.pa_cos[threadIdx.z][threadIdx.y] = in_parallactic_angle_cos[time * na + ant];
     }
 
     __syncthreads();
 
-    if(corr >= 4 || chan >= nchan || ant >= na || time >= ntime) return;
+    FT pa_sin = shared.pa_sin[threadIdx.z][threadIdx.y];
+    FT pa_cos = shared.pa_cos[threadIdx.z][threadIdx.y];
+
     for(int src = 0; src < nsrc; src++){
-        FT l = (in_coords[(src * 4 + corr) * 2 ] + shared.pointing_error[threadIdx.z][threadIdx.y][threadIdx.x>>2].x) * shared.antenna_scaling[threadIdx.y][threadIdx.x>>2].x; 
-        FT m = (in_coords[(src * 4 + corr) * 2 + 1] + shared.pointing_error[threadIdx.z][threadIdx.y][threadIdx.x>>2].y) * shared.antenna_scaling[threadIdx.y][threadIdx.x>>2].y; 
+        FT l = in_coords[src * 2]; 
+        FT m = in_coords[src * 2 + 1];
```

Comment:
`in_coords` should be a float2/double2 so that you can load the entire coordinate in in one memory operation. Use the appropriate type from abstraction.cuh